### ISSUE TITLE
Pool.Close() hangs indefinitely / Query method acquires connection, but does not return it?

### ISSUE THREAD

**Issue by michaeladler** [jackc/pgx#916]
Hi,

I've only recently started using pgx, so please forgive me if I'm doing something obviously wrong.

**Problem**

I'm trying to shut down `*pgxpool.Pool` cleanly via its `Close()` method

https://github.com/jackc/pgx/blob/210a217818576dcd21659594d289917b05ee7654/pgxpool/pool.go#L328

before I exit my application. This **blocks indefinitely** (or a very long time) and I have to finally `kill -9` my application.

**RCA**

Below are the connection stats before I invoke `Close()`:

```
acquireCount=3 acquireDuration=7.185483ms acquiredConns=1 canceledAcquireCount=0 constructingConns=0 emptyAcquireCount=2 idleConns=1 totalConns=2
```

Since `acquiredConns=1`,  `Close()` has to block and wait until the acquired connection is returned. The acquired connection is the result of `Pool.Query()`.

So my actual question is: Why does

https://github.com/jackc/pgx/blob/210a217818576dcd21659594d289917b05ee7654/pgxpool/pool.go#L423

`Acquire` a connection, **but not** `Release` it? If you compare this with `QueryFunc`

https://github.com/jackc/pgx/blob/210a217818576dcd21659594d289917b05ee7654/pgxpool/pool.go#L453

the connection is Released.

**Proposed Fix**:

Inserting `defer c.Release()` fixes the problem for me:

```go
func (p *Pool) Query(ctx context.Context, sql string, args ...interface{}) (pgx.Rows, error) {
	c, err := p.Acquire(ctx)
	if err != nil {
		return errRows{err: err}, err
	}
	defer c.Release()

	rows, err := c.Query(ctx, sql, args...)
	if err != nil {
		return errRows{err: err}, err
	}

	return c.getPoolRows(rows), nil
}
```

There are more places like this though, e.g. ` QueryRow`.

Should I do something differently or is this indeed a bug? If so, I'm happy to send a PR.

**Comment by jackc** [comment]
It should be released when the row is scanned.

**Comment by michaeladler** [comment]
Ah, I see. So the usage pattern of `Query()` is something like this:

```go
rows, err := pool.Query(...)
// do whatever I want
defer rows.Close()
```

I didn't `Close()` the rows, so that's the reason for the leaked connection.
Anyway, I've migrated my code to `QueryFunc()` so I don't have to worry about forgetting to close something.

Closing the issue here now since I don't think the behavior of `Query` can be improved. Maybe the documentation could be improved though :) Thanks!
